crates/circuit/src/parameter/parameter_expression.rs

```diff
@@ -0,0 +1,2162 @@
+        }
+    }
+
+    /// Clone the expression.
+    pub fn copy(&self) -> Self {
+        // TODO: Is this necessary? Doesn't ParameterExpression.clone() implement the same thing?
+        Self {
+            expr: self.expr.clone(),
+            name_map: self.name_map.clone(),
+        }
+    }
+
+    /// Complex conjugate the expression.
+    pub fn conjugate(&self) -> Self {
+        Self {
+            expr: self.expr.conjugate(),
+            name_map: self.name_map.clone(),
+        }
+    }
+
+    /// Compute the derivative of the expression with respect to the provided symbol.
+    ///
+    /// Note that this keeps the name map unchanged. Meaning that computing the derivative
+    /// of ``x`` will yield ``1`` but the expression still owns the symbol ``x``. This is
+    /// done such that we can still bind the value ``x`` in an automated process.
+    pub fn derivative(&self, param: &Symbol) -> Result<Self, String> {
+        Ok(Self {
+            expr: self.expr.derivative(param)?,
+            name_map: self.name_map.clone(),
+        })
+    }
+
+    /// Expand the expression.
+    pub fn expand(&self) -> Self {
+        Self {
+            expr: self.expr.expand(),
+            name_map: self.name_map.clone(),
+        }
+    }
+
+    /// Substitute symbols with [ParameterExpression]s.
+    ///
+    /// Args:
+    ///     - map: A hashmap with [Symbol] keys and [ParameterExpression]s to replace these
+    ///         symbols with.
+    ///     - allow_unknown_parameters: If ``false``, returns an error if any symbol in the
+    ///         hashmap is not present in the expression. If ``true``, unknown symbols are ignored.
+    ///         Setting to ``true`` is slightly faster as it does not involve additional checks.
+    ///
+    /// Returns:
+    ///     A parameter expression with the substituted symbols.
+    pub fn subs(
+        &self,
+        map: &HashMap<Symbol, Self>,
+        allow_unknown_parameters: bool,
+    ) -> Result<Self, ParameterError> {
+        // Build the outgoing name map. In the process we check for any duplicates.
+        let mut name_map: HashMap<String, Symbol> = HashMap::new();
+        let mut symbol_map: HashMap<Symbol, SymbolExpr> = HashMap::new();
+
+        // If we don't allow for unknown parameters, check if there are any.
+        if !allow_unknown_parameters {
+            let existing: HashSet<&Symbol> = self.name_map.values().collect();
+            let to_replace: HashSet<&Symbol> = map.keys().collect();
+
+            // This could be done a little bit more efficiently, but we want to
+            // show the user all symbols that are not present.
+            let difference: HashSet<Symbol> = to_replace
+                .difference(&existing)
+                .map(|&symbol| symbol.clone())
+                .collect();
+            if !difference.is_empty() {
+                return Err(ParameterError::UnknownParameters(difference));
+            }
```

Comment:
Can you defer the clone until we know we're in an error path?

```suggestion
            let difference: HashSet<&Symbol> = to_replace
                .difference(&existing)
                .collect();
            if !difference.is_empty() {
                return Err(ParameterError::UnknownParameters(difference.cloned()));
            }
```

(untested)

Reply:
That's a good point -- I changed it to the following trying also to avoid collecting the `difference` iterator. What do you think?

```rust
            let mut difference = to_replace.difference(&existing).peekable();

            if difference.peek().is_some() {
                let different_symbols = difference.map(|s| (**s).clone()).collect();
                return Err(ParameterError::UnknownParameters(different_symbols));
            }
```
I'm not very fond of the `**s` bit, but I guess this is the way?